KomaMRIFiles/src/Sequence/WritePulseq.jl

```diff
@@ -0,0 +1,407 @@
+    for ioamtd ∈ grad_idx_obj_amp_iamp_itim_delay
+        obj = ioamtd[2]
+        ioamtd[3] = γ * magsign(obj.A)
+        ioamtd[6] = round(1e6 * obj.delay)
+        shape_amp = obj.A / magsign(obj.A)
+        for (shape_amp_unique, id_amp) ∈ gradunique_amp_id
+            if length(shape_amp) == length(shape_amp_unique) && shape_amp ≈ shape_amp_unique
+                ioamtd[4] = id_amp
+            end
+        end
+        if isa(obj.T, Vector{<:Number})
+            shape_tim = cumsum([0; obj.T]) / seq.DEF["GradientRasterTime"]
+            for (shape_tim_unique, id_tim) ∈ gradunique_tim_id
+                if length(shape_tim) == length(shape_tim_unique) && shape_tim ≈ shape_tim_unique
+                    ioamtd[5] = id_tim
+                end
+            end
+        end
+    end
+
+    # Define the table to be written for the [TRAP] section
+    trap_idx_obj_amp_rise_flat_fall_delay = [[idx, obj, 0, 0, 0, 0, 0] for (obj, idx) ∈ trapunique_obj_id]
+    for ioarfad ∈ trap_idx_obj_amp_rise_flat_fall_delay
+        obj = ioarfad[2]
+        ioarfad[3] = γ * obj.A
+        ioarfad[4] = 1e6 * obj.rise
+        ioarfad[5] = 1e6 * obj.T
+        ioarfad[6] = 1e6 * obj.fall
+        ioarfad[7] = 1e6 * obj.delay
+    end
+
+    # Define the table to be written for the [ADC] section
+    adc_idx_obj_num_dwell_delay_freq_phase = [[idx, obj, 0, 0, 0, 0, 0] for (obj, idx) ∈ adcunique_obj_id]
+    for ionwdfp ∈ adc_idx_obj_num_dwell_delay_freq_phase
+        obj = ionwdfp[2]
+        ionwdfp[3] = obj.N
+        ionwdfp[4] = obj.T * 1e9 / (obj.N - 1)
+        ionwdfp[5] = (obj.delay - 0.5*obj.T/(obj.N - 1)) * 1e6
+        ionwdfp[6] = obj.Δf
+        ionwdfp[7] = obj.ϕ
+    end
+
+    # Define the table to be written for the [SHAPES] section
+    # Fix last sample of RF
+    [push!(a[1], a[1][end]) for a ∈ rfunique_abs_id]
+    [push!(a[1], a[1][end]) for a ∈ rfunique_ang_id]
+    shapefull_data_id = [shapeunique_data_id_i for shapeunique_data_id ∈ [rfunique_abs_id, rfunique_ang_id, rfunique_tim_id, gradunique_amp_id, gradunique_tim_id] for shapeunique_data_id_i ∈ shapeunique_data_id]
+    shape_data_id_num = [(length(compress(data)) == length(data) ? data : compress(data), id, length(data)) for (data, id) ∈ shapefull_data_id]
+
+    # Write the .seq file
+    open(filename, "w") do fid
+
+        @printf(fid, "# Pulseq sequence file\n")
+        @printf(fid, "# Created by KomaMRI.jl \n\n") #TODO: add Koma version
+
+        @printf(fid, "[VERSION]\n")
+        @printf(fid, "major 1\n")
+        @printf(fid, "minor 4\n")
+        @printf(fid, "revision 1\n")
+        @printf(fid, "\n")
+
+        if !isempty(seq.DEF)
+            @printf(fid, "[DEFINITIONS]\n")
+            sorted_keys = sort(collect(keys(seq.DEF)))
+            for key ∈ sorted_keys
+                val = seq.DEF[key]
+                if key ∈ ["AdcRasterTime", "BlockDurationRaster", "GradientRasterTime", "RadiofrequencyRasterTime", "TotalDuration", "FOV", "MaxAdcSegmentLength"]
```

Comment:
Why only these? Check the file I sent you

Reply:
Now is writting all the definitions. Addressed in 08cd263.